[CORE] Successive and combined project negotiations

stefaus opened the pull request:
Currently Saros creates a new project negotiation for every individually
added project resource. This leads to multiple concurrent negotiations
per user, which creates more overhead for transportation and user
interaction for every single negotiation. Additionally, it leads to
unexpected behavior because a new negotiation will transfer all files
currently not available on the remote machine, including files from
unfinished negotiation transfers.

This patch ensures that only one project negotiation per user is running
and all new shared resources are queued till all ongoing negotations
finished, before starting a new negotiation with all queued resources /
projects combined.

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -532,44 +550,105 @@ public void addResourcesToSession(
+                synchronized (nextSharedResourcesWorkerNotify) {
+                    nextSharedResourcesWorkerNotify.wait();
+                }
+            } catch (InterruptedException e) {
```
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: Either re-interrupt this method or rethrow the "InterruptedException". ([squid:S2142](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS2142))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWaT0XAZyONdBoB89fGy&open=AWaT0XAZyONdBoB89fGy)

[ISSUE_KEY]: AWaT0XAZyONdBoB89fGy

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -135,6 +146,13 @@ public void negotiationTerminated(final SessionNegotiation negotiation) {
+            if (!currentProjectNegotiations.negotiationsInProgress()) {
+                /** notify about the completion of all running negotiations **/
+                synchronized (nextSharedResourcesWorkerNotify) {
+                    nextSharedResourcesWorkerNotify.notify();
```
![Bug](https://sonarsource.github.io/sonar-github/bug.png 'Bug') Bug: "notify" may not wake up the appropriate thread. ([squid:S2446](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS2446))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWaT0XAZyONdBoB89fGw&open=AWaT0XAZyONdBoB89fGw)

[ISSUE_KEY]: AWaT0XAZyONdBoB89fGw

srossbach commented:
> Additionally, it leads to
unexpected behavior because a new negotiation will transfer all files
currently not available on the remote machine, including files from
unfinished negotiation transfers.

Please fix this issue first. We can talk about enhancements later. But I do not see any point of improving the current system. You rarely share anything additional during the session, you can share multiple projects with the start of a session. Furthermore you cannot just change the logic of some methods to a blocking state. At least I have seen some issues when at least some user do not accept the negotiation (eg. gone AFK). You simply block anything for every user.

srossbach commented:
1. 95% of the use cases are = full sharing
2. Starting multiple project negotiation with redundant data seems to be a bug in the SessionManager (the fix should be trivial)
3, We have a wizard / dialog where you can select multiple resources that should be shared. The better fix is to disable this right click an share stuff.

Drakulix commented:
>     3. We have a wizard / dialog where you can select multiple resources that should be shared. The better fix is to disable this right click an share stuff.

Yes that would be a better fix, if the only intend of this PR would be fixing this bug.

But as said this simplifies the Instant Session Start implementation as well as my upcoming PR for non-host project sharing, which is needed for Saros Server.

@fzieris was also in favor of this change, as allowing multiple `ProjectNegotiation`s in parallel is not a very useful feature in the vast majority of cases.

This new limitation will simplify future code, addresses an existing bug - which does not mean we cannot improve the logic of the `SessionManager` anyway - and does not need to change the UI as with your suggestion.

I would argue changing the UI is a separate issue, which can be discussed independently of the PR, if you view the "right-click and share" as problematic. I personally think a change in the user experience is much more problematic and we should not need to do that, if we can still easily support the old UI with this PR.

That said I have not yet had the time to actually review the code, so this is not a statement over the overall code quality just on the issue of "not seeing any point of improving the current system".

stefaus commented:
> 1. 95% of the use cases are = full sharing

Yes we talk about a maybe rarely used feature, so the user impact should be low. 

> 2. Starting multiple project negotiation with redundant data seems to be a bug in the SessionManager (the fix should be trivial)

From my point of view: The problem for this case lies in the invitation logic, because a negotiation is trying to synchronize all project resources added to the session.
In `SarosSessionManager.addResourcesToSession(Map<IProject, List<IResource>> projectResourcesMapping)` new projects/resources are added to the Session and the belonging Project(s) marked for a Project Negotiation (`projectsToShare`)
https://github.com/saros-project/saros/blob/67f866410eb4318fa660c3077a34fd0b79c96fec/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java#L551-L554

than get handed to a new Project Negotation
https://github.com/saros-project/saros/blob/67f866410eb4318fa660c3077a34fd0b79c96fec/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java#L584-L586

in `AbstractOutgoingProjectNegotiation` the list of projects is used (without naming the resources of this negotiation) to create a file list
https://github.com/saros-project/saros/blob/67f866410eb4318fa660c3077a34fd0b79c96fec/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/AbstractOutgoingProjectNegotiation.java#L95-L96

`createProjectNegotiationDataList(...)` uses all the files added to the session in `SessionManager`
https://github.com/saros-project/saros/blob/67f866410eb4318fa660c3077a34fd0b79c96fec/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/AbstractOutgoingProjectNegotiation.java#L340-L344

If this would be called just with the added resources list from `SessionManager`, the diff part of the invitation logic would probably result in deleting all other (remote existing) project files.

> 3. We have a wizard / dialog where you can select multiple resources that should be shared. The better fix is to disable this right click an share stuff.

The Wizard allows the same use.

srossbach commented:
Well for partial sharing removed resources are discarded, see https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/FileListDiff.java#L61

Ok so this is my bad and the fix ain't that trivial. However I would recommend to move the aggregation logic to a class. The SarosSessionManager code is already complex enough. Adding more synchronization objects do not make it easier to understand. Furthermore the ISarosSession interface needs a Javadoc update. If you call this method from the GUI thread you can freeze the whole GUI.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ProjectNegotiationObservable.java:
```diff
@@ -116,4 +116,15 @@ public synchronized void remove(ProjectNegotiation negotiation) {
+    /**
+     * @return true if running negotiations exist
+     */
+    public synchronized boolean negotiationsInProgress() {
```
I would give this a better name or drop it completely as calling isEmpty on the result returned by list() yields the same result.

stefaus replied:
any name suggestion?
No!, calling isEmpty on the result of list() yields to a different result (you can / will get a list of maybe empty lists) and list() is creating a new ArrayList for every call.

stefaus replied:
Shame on me, Sorry, I mixed it up with calling `negotiations.values().isEmpty()`.

Ok, than I change it to `isEmpty()` and replace other occurrences.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/ProjectNegotiationObservable.java:
```diff
@@ -116,4 +116,15 @@ public synchronized void remove(ProjectNegotiation negotiation) {
     }
+
+    /**
+     * @return true if running negotiations exist
```
Yeah I know the method description will be the same as the return description. However, just add it.

tobous replied:
The official oracle doc guide states this from the reversed situation, but the outcome is the same:
> include it for all other methods, even if its content is entirely redundant with the method description

https://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@return

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -114,6 +115,16 @@
 
     private final ProjectNegotiationObservable currentProjectNegotiations;
 
+    /** to queue resources for the next project negotiation **/
```
That is what the HashMap is supposed to be used for. The description is not really helpful.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -114,6 +115,16 @@
 
+    /** to queue resources for the next project negotiation **/
+    private final Map<IProject, List<IResource>> nextSharedResources = new HashMap<IProject, List<IResource>>();
+    /** to set the thread for handling the next project negotiation **/
```
I had a hard time to understand this. Basically it is a flag / mutex ... better a semaphore

stefaus replied:
It is a semaphore as a plain variable. Will add something to the doc.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -114,6 +115,16 @@
+    /** to set the thread for handling the next project negotiation **/
+    private volatile boolean nextSharedResourcesWorker = false;
+    /**
+     * to notify a thread waiting for completion of all running project
```
See my comment before this review, move the code to a dedicated class. The comments aint that helpful, even if you look a the code where all these objects are used.

stefaus replied:
I would prefer to keep it at this place for now and change it with the refactoring of the package.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -135,6 +146,13 @@ public void negotiationTerminated(final SessionNegotiation negotiation) {
             currentProjectNegotiations.remove(negotiation);
+
+            if (!currentProjectNegotiations.negotiationsInProgress()) {
+                /** notify about the completion of all running negotiations **/
```
Do not repeat the code in your comments.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -532,44 +550,104 @@ public void addResourcesToSession(
+        synchronized (this) {
+            for (Entry<IProject, List<IResource>> mapEntry : projectResourcesMapping
+                .entrySet()) {
+                final IProject project = mapEntry.getKey();
```
if (!containsKey(project) map.put(project, new Set<>)

if (resourceList == null || resourceList.isEmpty) {map.put(project, fullSharedMarkerSet); return;}

Set set = map.get(project)

if set ==  fullSharedMarkerSet
  return;

set.addAll(resourcesList)

Should be easier to understand.

stefaus replied:
Please take in account that a new empty Set creates a partial project without resources.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -532,44 +550,104 @@ public void addResourcesToSession(
+                try {
+                    nextSharedResourcesWorkerNotify.wait();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
```
return ?

stefaus replied:
would break the logic, only case escaping this would be a ending session or not?

srossbach replied:
final Object l = new Object();
		int x = 0;
		
		Thread.currentThread().interrupt();
		
		synchronized(l) {
			while(x < 1000) {
				try {
					l.wait();
				} catch (InterruptedException e) {
					System.err.println("interrupted");
					Thread.currentThread().interrupt();
					x++;
				}
			}
		}
translate into busy spinning

If you want to ignore the interruption then catch the exception, remember that you got interrupted and when your logic is done call Thread#interrupt();

However getting interrupted is a signal to stop the thread gracefully, and so it should terminate as early as it can.

srossbach commented:
@tobous I added you to the review list as changes in the core can have in impact on some IntelliJ stuff as well. So it is not necessary to review the code but the impact on the changes it might produce.

But I guess there are not any ?

tobous commented:
@srossbach I don't know. I haven't come around to dealing with/looking at that stuff in the IntelliJ part yet.

But for now, I don't think this should have any relevance for the IntelliJ part as currently participants are only able to share a single module. There is no way of sharing multiple modules or to add resources to an existing session. Furthermore, sessions can only be created with one participant (+ host) and there is also no way of adding participants. All of this is restricted by the UI, but I don't know if there are any other obstacles (or if this is all done in the core and only the UI is missing).

I tested the commit locally and the project negotiation still seems to work as before. Do you know any any general part of the logic in the IntelliJ part (or specific classes; except the ui stuff) that this could affect?

srossbach commented:
Stefans current solution has added blocking behavior. If you do not call the affected method from the Application thread you have not to worry about anything.

As for adding users to the session and also adding more resources see: https://github.com/saros-project/saros/blob/master/de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/core/ui/util/CollaborationUtils.java

Looking at the code you are already on the safe side.

If I am not mistaken adding a user should work without any issues. However I always suggest to start the "induction" proof with n=2 users and the step from n=2 to n=3 . We have already seen some solutions that work fine in an initial 2 user setup but then fail if another user joins the session.

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -532,66 +588,43 @@ public void addResourcesToSession(
+                    .getUserProperties(user).getString(
+                        ProjectNegotiationTypeHook.KEY_TYPE));
+                AbstractOutgoingProjectNegotiation negotiation = negotiationFactory
+                    .newOutgoingProjectNegotiation(user.getJID(), type,
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this use of "getJID"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWbfy2AyLpqqSUlcop4G&open=AWbfy2AyLpqqSUlcop4G)

[ISSUE_KEY]: AWbfy2AyLpqqSUlcop4G

saros-infrastructure replied:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this use of "getJID"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWbfy2AyLpqqSUlcop4G&open=AWbfy2AyLpqqSUlcop4G)

[ISSUE_KEY]: AWbfy2AyLpqqSUlcop4G

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/ProjectNegotiationCollector.java:
```diff
@@ -0,0 +1,126 @@
+     * @param projectResourcesMapping
+     *            project resource mappings to add
+     */
+    public synchronized void add(
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([squid:S3776](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWbfy2BCLpqqSUlcop4H&open=AWbfy2BCLpqqSUlcop4H)

[ISSUE_KEY]: AWbfy2BCLpqqSUlcop4H

saros-infrastructure replied:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([squid:S3776](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWbfy2BCLpqqSUlcop4H&open=AWbfy2BCLpqqSUlcop4H)

[ISSUE_KEY]: AWbfy2BCLpqqSUlcop4H

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -744,7 +777,7 @@ private boolean terminateNegotiations() {
 
         while (System.currentTimeMillis() - startTime < NEGOTIATION_TIMEOUT) {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Reduce the total number of break and continue statements in this loop to use at most one. ([squid:S135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWbfy2AyLpqqSUlcop4E&open=AWbfy2AyLpqqSUlcop4E)

[ISSUE_KEY]: AWbfy2AyLpqqSUlcop4E

saros-infrastructure replied:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Reduce the total number of break and continue statements in this loop to use at most one. ([squid:S135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWbfy2AyLpqqSUlcop4E&open=AWbfy2AyLpqqSUlcop4E)

[ISSUE_KEY]: AWbfy2AyLpqqSUlcop4E

stefaus commented:
Sorry for the delay.

Now I moved logic to a new class and removed the blocking behaviour (with a new thread if needed).

@tobous My goal is to replan the whole invitation process and make it more user friendly, like uniting separate invitation dialogs and delays between them.
In addition refactor this core part and hopefully make IntelliJ and the new GUI Implementaion easier. No fear I will talk to you about it, make a plan and present it in a ticket before work will start.

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/ProjectNegotiationCollector.java:
```diff
@@ -0,0 +1,123 @@
+     */
+    public synchronized void add(
+        Map<IProject, List<IResource>> projectResourcesMapping) {
+        for (Entry<IProject, List<IResource>> mapEntry : projectResourcesMapping
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Reduce the total number of break and continue statements in this loop to use at most one. ([squid:S135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWb1y62KW3Y69bz5Xor5&open=AWb1y62KW3Y69bz5Xor5)

[ISSUE_KEY]: AWb1y62KW3Y69bz5Xor5

stefaus commented:
@srossbach Could you please review again, I hope all points are addressed.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/ProjectNegotiationCollector.java:
```diff
@@ -0,0 +1,122 @@
+     * 
+     * @return the list of projects to share
+     */
+    public synchronized List<IProject> getProjectsToShare(
```
The name of this method does not reflect what it really does. Accessing the session should be done in the SessionManager and not here. The collector here should mainly merge resources that are required when for the next session negotiation.

stefaus replied:
I moved the code back in the manager.

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -561,37 +616,34 @@ public void addResourcesToSession(
+                    .getUserProperties(user).getString(
+                        ProjectNegotiationTypeHook.KEY_TYPE));
+                AbstractOutgoingProjectNegotiation negotiation = negotiationFactory
+                    .newOutgoingProjectNegotiation(user.getJID(), type,
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this use of "getJID"; it is deprecated. ([squid:CallToDeprecatedMethod](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ACallToDeprecatedMethod))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWb814lgTJjFgJLT72lr&open=AWb814lgTJjFgJLT72lr)

[ISSUE_KEY]: AWb814lgTJjFgJLT72lr

saros-infrastructure commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -744,7 +796,7 @@ private boolean terminateNegotiations() {
 
         while (System.currentTimeMillis() - startTime < NEGOTIATION_TIMEOUT) {
```
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Reduce the total number of break and continue statements in this loop to use at most one. ([squid:S135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=272&issues=AWb814lgTJjFgJLT72lp&open=AWb814lgTJjFgJLT72lp)

[ISSUE_KEY]: AWb814lgTJjFgJLT72lp

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -532,66 +588,43 @@ public void addResourcesToSession(
-                log.warn("could not start a project negotiation because the current session is about to stop");
-                return;
-            }
+        if (!startStopSessionLock.tryLock()) {
```
Ok now I remember my logic. I used synchronized as second lock to avoid a possible situation where adding project and receiving a project concurrently would result in one of the operation being discarded. However this is currently not possible due to the architecture (only the host can add projects and only client can receive projects). We still have a patch from Denis that try to reintroduce this behavior for the Saros server. So I do not think moving the sychronized block is the best option here as you increase the size of the code that is being locked.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -135,6 +139,12 @@ public void negotiationTerminated(final SessionNegotiation negotiation) {
             currentProjectNegotiations.remove(negotiation);
+
+            if (currentProjectNegotiations.isEmpty()) {
+                synchronized (nextProjectNegotiation) {
```
Why isn't this behavior moved to the class instead, e.g creating a signal method. The code is getting too complicated. If I count correctly there a now 3 locks involved for the logic. I think we should prefer an asynchronous solution instead of a concurrent one, e.g:

collector.add(resources)
excecutor.submit(runSharing)

the runSharing runnable can then look if there is a running negotiation, if yes abort. if there are no resources, abort

This is somewhat easier to handle, e.g you can execute the code 1 time or 1000 times it will not matter, the result be the same.

Drakulix commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/ProjectNegotiationCollector.java:
```diff
@@ -0,0 +1,96 @@
+ * This class allows to hold the reference to a thread that should handle this
+ * new project negotiation.
+ */
+public class ProjectNegotiationCollector {
```
Why is this class holding the worker thread?

It collects mappings just fine, the worker thread stuff seems tagged on. Both parts could be separated easily and I do not think the "Collector" is responsible for the worker thread, but rather the SessionManager, which even creates the worker.

I get that this needs to be synchronized with the other methods, but responsibilities are a bit unclear and it would be nice, if this could be cleaned up a bit, as this stuff belongs into the SessionManager.

stefaus replied:
My thought was to group logic to enable calls like `nextProjectNegotiation.isWorkerAlive()`, but rethinking it now... you're right and I moved it.

Drakulix commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -533,20 +589,19 @@ public void addResourcesToSession(
-        for (Entry<IProject, List<IResource>> mapEntry : projectResourcesMapping
-            .entrySet()) {
-
+        for (Entry<IProject, List<IResource>> mapEntry : nextProjectNegotiation
```
A little comment on the locking logic would not hurt. I have to check the get-function to figure out that this loop is locking. Same applies for the isEmpty call. You could make it a bit more obvious what is going on, so it is easier to check, if any of this could trigger a deadlock.

stefaus replied:
I moved it to a separate variable, hope that helps.

In general I'm not sure what you think I should comment, for example exisiting calls like `currentProjectNegotiations.add(negotiation);` are blocking as well without further comment.

srossbach commented on de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java:
```diff
@@ -485,8 +497,56 @@ public void invite(Collection<JID> jidsToInvite, String description) {
+   * This method handles new project negotiations for already invited user (not the first in the
+   * process of inviting to the session).
+   */
+  private synchronized void startNextProjectNegotiation() {
```
Please pull the synchronized down to line 603 so it matches the former synchronized block. It should be clear what section must be synchronized. Synchronizing the whole method is not very helpful for understanding.
